### PR TITLE
Add proxy config to rest client without intercept

### PR DESCRIPTION
**Pull request by shaigr**
## Summary
Added proxy configuration for BaseRestClient to allow going through a proxy server
new configuration added: proxy_host, proxy_port, proxy_protocol
## Additional Information
old behavior will not be affected if no proxy configuration is supplied
## Testing
Tested and verified with caddy reverse proxy, I was able to send order to bybit through it.

### CODE REVIEW

File: src/util/BaseRestClient.ts [hunk 1]
```diff
@@ -122,6 +124,16 @@ export default abstract class BaseRestClient {
...
 
+    if (this.options.proxy_host){
+      this.proxy = {
+        protocol: this.options.proxy_protocol!,
+        host: this.options.proxy_host,
+        port: this.options.proxy_port!
+      };
+    } else {
+      this.proxy = undefined;
+    }
```
Comment (tiagosiebler) [inline src/util/BaseRestClient.ts:135]: Thanks for the PR & sorry for the delay. It's actually possible for you to inject these axios-parameters without any code change in the BaseRestClient. Take a look at the second parameter in the constructor of the BaseRestClient (line 108). `networkOptions` is of the type `AxiosRequestConfig`. This means you can set your axios-compatible configuration when making an instance of the rest client you're using. Example:
```typescript
const key = process.env.API_KEY_COM;
const secret = process.env.API_SECRET_COM;

const restClientOpts: RestClientOptions = {
  key: key,
  secret: secret,
};

const axiosOpts: AxiosRequestConfig = {
  proxy: {
    protocol: 'https',
    host: 'yourproxyhost',
    port: 5432
  }
};

const client = new RestClientV5(restClientOpts, axiosOpts);
```

No code change needed to support this, already works with the latest version of all my SDKs.

Comment (shaigr) [reply src/util/BaseRestClient.ts:135]: I feel so silly for missing the option to pass these axiosOpts to the constructor..
Thank you for the review and example!

Comment (tiagosiebler) [reply src/util/BaseRestClient.ts:135]: Don't worry at all, it's probably not obvious enough and none of the examples use it! Either way I appreciate that you took the time to propose a PR to meet your goal - most people just ask and don't try to implement it, so I hugely appreciate your initiative 🙂

Comment (shaigr) [reply src/util/BaseRestClient.ts:135]: While we are at it - have you ever tried to use a proxy for websockets (isomorphic-ws)?
I noticed it was possible to pass to bybit-api's websocketClient an agent in its requestOptions.
The two options I tried to pass for the agent are:
* https://www.npmjs.com/package/https-proxy-agent
* https://www.npmjs.com/package/better-https-proxy-agent
I tried both https/http protocol for the proxy agent.

I've tried multiple configurations for the proxy servers:
* nginx
* socat (simple tcp tunneling, listening on https)
* websocat (https://github.com/vi/websocat)

with no luck, I think most of the errors I ran into had to do with SSL handshakes failing..

Currently I've managed to get it working only in one way -
1. Using websocat as the proxy server (listening on ws:// and forwarding to wss:// on bybit)
2. it doesn't work with any of the above proxy agents, but it works if I patch the bybit-api to send requests to the proxy server on ws://

`     linear: {
         public: {
            livenet: 'ws://127.0.0.1:8082/realtime_public',
          livenet2: 'ws://127.0.0.1:8082/realtime_public',
            testnet: 'ws://127.0.0.1:8082/realtime_public',
`

Comment (tiagosiebler) [reply src/util/BaseRestClient.ts:135]: I have gotten this working on the binance SDK, through the 3rd parameter of the WebSocket constructor:
https://github.com/tiagosiebler/binance/blob/master/src/websocket-client.ts#L227-L229

```
  const httpsAgent = getHttpsProxyAgent(
    restClientProxy.host,
    restClientProxy.port,
    restClientProxy.auth.username,
    restClientProxy.auth.password,
  );

  const wsClient = new WebsocketClient(
    {
      api_key: key,
      api_secret: secret,
      beautify: true,
      requestOptions: {
        // httpsAgent: httpsAgent,
        proxy: restClientProxy,
      },
      wsOptions: {
        agent: httpsAgent,
      },
    },
    logger,
  );
  ```
  
  Seems there's a slight inconsistency now between my bybit SDK and the binance one:
  https://github.com/tiagosiebler/bybit-api/blob/master/src/websocket-client.ts#L960
  
  You should be able to prepare the https proxy agent the same way (using https-proxy-agent), and then pass it through the requestOptions part of the ws client config:
  
```
  const httpsAgent = new HttpsProxyAgent(url);
  const wsClient = new WebsocketClient(
  {
    // key: key,
    // secret: secret,
    market: 'v5',
      requestOptions: {
        // httpsAgent: httpsAgent,
        agent: httpsAgent,
      },
  },
  logger
);


```

Based on the implementation, I think that should work without code change in the SDK!

Comment (tiagosiebler) [reply src/util/BaseRestClient.ts:135]: Oh and here's how the getHttpsProxyAgent uses that https-proxy-agent npm module: https://github.com/tiagosiebler/binance/blob/master/src/util/proxy.ts#L6

Comment (shaigr) [reply src/util/BaseRestClient.ts:135]: Thanks! That is exactly what I've tried.. may I ask which tool did you use for the proxy server?

Comment (tiagosiebler) [reply src/util/BaseRestClient.ts:135]: In this case it was with brightdata (used to be called luminati). Haven't tried hosting my own proxy in a long time...would be curious though about tools for it.

Comment (shaigr) [reply src/util/BaseRestClient.ts:135]: Interesting, thanks for sharing that. As I said, I was am trying to use socat / websocat / nginx as hosted proxy servers, which work for me just fine when I use them as proxy for REST API but I had no luck getting them to work with web sockets by passing the agent parameter in request options. I keep getting SSL handshake issues. 

**However**, I reviewed the code some more, and I see that **I can still manage without patching the bybit-api code**. I can just override the wsUrl parameter, in the options for the websocketClient constructor. That'll allow me to use the working solution I mentioned in my previous comment (without using the https proxy agent).

**Pull request closed without merging**